serialization/parse_node_factory_registry.go

```diff
@@ -1,45 +1,57 @@
-package serialization
-
-import (
-	"errors"
-	re "regexp"
-	"strings"
-)
-
-// ParseNodeFactoryRegistry holds a list of all the registered factories for the various types of nodes.
-type ParseNodeFactoryRegistry struct {
-	ContentTypeAssociatedFactories map[string]ParseNodeFactory
-}
-
-// DefaultParseNodeFactoryInstance is the default singleton instance of the registry to be used when registering new factories that should be available by default.
-var DefaultParseNodeFactoryInstance = &ParseNodeFactoryRegistry{
-	ContentTypeAssociatedFactories: make(map[string]ParseNodeFactory),
-}
-
-// GetValidContentType returns the valid content type for the ParseNodeFactoryRegistry
-func (m *ParseNodeFactoryRegistry) GetValidContentType() (string, error) {
-	return "", errors.New("the registry supports multiple content types. Get the registered factory instead")
-}
-
-var contentTypeVendorCleanupPattern = re.MustCompile("[^/]+\\+")
-
-// GetRootParseNode returns a new ParseNode instance that is the root of the content
-func (m *ParseNodeFactoryRegistry) GetRootParseNode(contentType string, content []byte) (ParseNode, error) {
-	if contentType == "" {
-		return nil, errors.New("contentType is required")
-	}
-	if content == nil {
-		return nil, errors.New("content is required")
-	}
-	vendorSpecificContentType := strings.Split(contentType, ";")[0]
-	factory, ok := m.ContentTypeAssociatedFactories[vendorSpecificContentType]
-	if ok {
-		return factory.GetRootParseNode(vendorSpecificContentType, content)
-	}
-	cleanedContentType := contentTypeVendorCleanupPattern.ReplaceAllString(vendorSpecificContentType, "")
-	factory, ok = m.ContentTypeAssociatedFactories[cleanedContentType]
-	if ok {
-		return factory.GetRootParseNode(cleanedContentType, content)
-	}
-	return nil, errors.New("content type " + cleanedContentType + " does not have a factory registered to be parsed")
-}
+package serialization
+
+import (
+	"errors"
+	re "regexp"
+	"strings"
+	"sync"
+)
+
+// ParseNodeFactoryRegistry holds a list of all the registered factories for the various types of nodes.
+type ParseNodeFactoryRegistry struct {
+	// Lock should be used when accessing other fields of this struct to ensure thread safety.
+	Lock *sync.Mutex
```

Comment:
could we make the lock private?

Reply:
Doing so would require exposing it indirectly via a public method. `api_client_builder.go` is not in the `serialization` package.` I can do that, what would the benefit be? Exposing a `Lock` and `Unlock` method directly on the struct (making it a `sync.Locker`) would prevent "outside" code from changing the field.

Reply:
Thanks for the additional context. I was a bit worried that'd be the case. The only benefit of hiding the lock behind methods would be to hide the lock type. Which I don't see as a huge benefit and which might come in the way of static analysis tools. No action required, resolving.

Reply:
we'll keep the wrapping method. Thanks for making the change.